### PR TITLE
extend mp_arith bitwise ops with checks for operands sizes/values

### PR DESCRIPTION
**Pull request by mgudemann**

### CODE REVIEW

File: src/util/mp_arith.cpp [hunk 1]
```diff
@@ -315,3 +319,253 @@ unsigned integer2unsigned(const mp_integer &n)
...
+ Purpose: bitwise or
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer bitwise_or(const mp_integer &a, const mp_integer &b)
+{
+  ullong_t result=a.to_ulong()|b.to_ulong();
```
Comment (tautschnig) [inline src/util/mp_arith.cpp]: Two questions: 1) what is the use case for those functions? 2) Shouldn't all of them check whether the value can be fully represented in a ullong_t?

File: src/util/mp_arith.cpp [hunk 2]
```diff
@@ -315,3 +319,253 @@ unsigned integer2unsigned(const mp_integer &n)
...
+ Outputs:
+
+ Purpose: left shift
+          bitwise operations only make sense on native objects, hence the
+          largest object size should be the largest available c++ integer
+          size (currently long long)
+
+\*******************************************************************/
+
+mp_integer left_shift(
```
Comment (tautschnig) [inline src/util/mp_arith.cpp]: What about the existing `operator<<`/`operator>>`?

File: src/util/mp_arith.cpp (hunk 1 above)
Comment (mgudemann) [inline src/util/mp_arith.cpp]: there's an internal use for them in the interpreter, where they will be applied to C/Java standard types only

File: src/util/mp_arith.cpp (hunk 2 above)
Comment (mgudemann) [inline src/util/mp_arith.cpp]: as above, there's bitwise operations for internal use by the interpreter

File: src/util/mp_arith.cpp (hunk 1 above)
Comment (tautschnig) [inline src/util/mp_arith.cpp]: Ok! So how about the proposed check as to whether the application is actually sound?

File: src/util/mp_arith.cpp (hunk 2 above)
Comment (tautschnig) [inline src/util/mp_arith.cpp]: I'm not sure I fully understand this one: the newly proposed `left_shift` vs `operator<<` have some overlap in what they are trying to achieve. When would one use the former, when the latter?

File: src/util/mp_arith.cpp (hunk 1 above)
Comment (mgudemann) [inline src/util/mp_arith.cpp]: I added the checks for the bitwise ops

File: src/util/mp_arith.cpp (hunk 2 above)
Comment (mgudemann) [inline src/util/mp_arith.cpp]: I simplified shift left, this version effectively uses `<<` but truncates to zero in case of overflow wrt. `true_size`

**Comment by kroening** [comment]
This isn't mergeable as-is; it's clear that somebody will fall for the restriction to ulong-sized mp_integers. If these are really needed, then fix please.

**Comment by mgudemann** [comment]
I added an `.is_ulong()` assert to each operand, which asserts that the number of used digits does not exceed what fits into `ullong_t`

**Comment by mgudemann** [comment]
@tautschnig @peterschrammel this is simplified now, as the bitwise operations have been merged from another PR meanwhile.

**Comment by tautschnig** [comment]
I suppose this should now be renamed? It does look good to me as-is.

**Comment by mgudemann** [comment]
@tautschnig I renamed the PR, I could rename the branch, too, but then I think I'll have to create a new PR

**Comment by tautschnig** [comment]
@mgudemann Thanks for the renaming of the PR and my apologies for having been unclear: I was not expecting a rename of the branch.

**Pull request merged**